### ISSUE TITLE
Question about Triplet Loss

### ISSUE THREAD

**Issue by wmpscc** [dzld00/pytorch-manifold-matching#1]
Thanks for your fantasy work. I have a question about Triplet Loss.
``` python
    def forward(self, anchor, positive, negative, size_average=True):
        distance_positive = (anchor - positive).pow(2).sum(1)  # .pow(.5)
        distance_negative = (anchor - negative).pow(2).sum(1)  # .pow(.5)
        
        cos = nn.CosineSimilarity(dim=1, eps=1e-6)
        cos_reg = cos(negative - anchor, positive - anchor).sum(0) 
        losses = F.relu(distance_positive - distance_negative + self.margin - self.alpha * cos_reg) #2e-2

        return losses.mean()
```

When the value of `distance_negative`  is higher than the value of `distance_positive`, the `losses` may be zeros. how to avoid it.

**Comment by dzld00** [comment]
One way is to play with parameters for example using a larger margin or smaller alpha, especially when using larger batch size.

**Comment by wmpscc** [comment]
@dzld00 
Thank you for your reply. But I still have another question: Why should the negative part of the loss be removed? What is the impact of not using relu?

**Comment by dzld00** [comment]
Thanks for the insightful question. The relu part should not impact a successful training session where the loss remains positive, as you said. It is mainly for aligning with the original design of Triplet loss.
